File: ts/src/models/event.ts
import {PlainValue} from "@code0-tech/tucana/helpers";
import {HerculesTranslation} from "../types";
import {FlowTypeSetting_UniquenessScope} from "@code0-tech/tucana/shared";
import {RuntimeEventClass} from "./runtime-event";

export interface HerculesEventSetting {
    identifier: string,
    unique?: FlowTypeSetting_UniquenessScope,
    linkedDataTypeIdentifiers?: string[],
    defaultValue?: PlainValue,
    name?: HerculesTranslation[],
    description?: HerculesTranslation[],
    optional?: boolean,
    hidden?: boolean,
}

export interface HerculesEvent {
    identifier: string,
    settings?: HerculesEventSetting[],
    signature: string,
    linkedDataTypes?: string[],
    editable?: boolean,
    name?: HerculesTranslation[],
    description?: HerculesTranslation[],
    documentation?: HerculesTranslation[],
    displayMessage?: HerculesTranslation[],
    alias?: HerculesTranslation[],
    displayIcon?: string,
    runtimeIdentifier?: string,
}

export interface EventDefinitionRunnable {}

export type EventClass<T extends RuntimeEventClass = RuntimeEventClass> = new () => InstanceType<T>;

File: py/hercules/models/event.py
"""Event (flow type) models (port of ``src/models/event.model.ts``)."""
from __future__ import annotations

from dataclasses import dataclass
from typing import List, Optional, Union

from hercules._tucana.helpers import PlainValue
from hercules.types import FlowTypeSetting_UniquenessScope, Translation


@dataclass
class EventSettingProps:
    identifier: str
    unique: Optional[Union[int, str]] = None
    linked_data_type_identifiers: Optional[List[str]] = None
    default_value: Optional[PlainValue] = None
    name: Optional[List[Translation]] = None
    description: Optional[List[Translation]] = None
    optional: Optional[bool] = None
    hidden: Optional[bool] = None


@dataclass
class EventModel:
    identifier: str
    signature: str
    settings: Optional[List[EventSettingProps]] = None
    editable: Optional[bool] = None
    name: Optional[List[Translation]] = None
    description: Optional[List[Translation]] = None
    documentation: Optional[List[Translation]] = None
    display_message: Optional[List[Translation]] = None
    alias: Optional[List[Translation]] = None
    display_icon: Optional[str] = None
    runtime_identifier: Optional[str] = None
